app/controllers/manageBusinesses/manage/CheckYourAnswersController.scala

```diff
@@ -0,0 +1,226 @@
+      val incomeSourceIdStringOpt = sessionData.manageIncomeSourceData.flatMap(_.incomeSourceId)
+      val chnageToStringOpt = sessionData.manageIncomeSourceData.flatMap(_.reportingMethod)
+      val taxYearStringOpt = sessionData.manageIncomeSourceData.flatMap(_.taxYear)
+      val incomeSourceIdOpt = incomeSourceIdStringOpt.map(id => IncomeSourceId(id))
+      handleShowRequest(taxYearStringOpt,
+        chnageToStringOpt,
+        isAgent,
+        incomeSourceType,
+        incomeSourceIdOpt,
+        backUrl = {
+          if (isAgent) controllers.routes.HomeController.showAgent
+          else controllers.routes.HomeController.show()
+        }.url
+      )
+    }
+  }
+
+  private def handleShowRequest(taxYearStringOpt: Option[Int], chnageToStringOpt: Option[String], isAgent: Boolean, incomeSourceType: IncomeSourceType, incomeSourceIdOpt: Option[IncomeSourceId], backUrl: String)(implicit user: MtdItUser[_]): Future[Result] = {
+    val maybeIncomeSourceId: Option[IncomeSourceId] = user.incomeSources.getIncomeSourceId(incomeSourceType, incomeSourceIdOpt.map(m => m.value))
+    withIncomeSourcesFS {
+      Future.successful(
+        (taxYearStringOpt, chnageToStringOpt, maybeIncomeSourceId) match {
+          case (Some(taxYearStringOpt), Some(changeToStringOpt), Some(id)) => {
+            val taxYearStartYear = taxYearStringOpt - 1
+            val taxYearEndYear = taxYearStringOpt
+            Ok(checkYourAnswers(
+              isAgent,
+              backUrl,
+              CheckYourAnswersViewModel(id, changeToStringOpt, taxYearStartYear.toString, taxYearEndYear.toString, incomeSourceType),
+              incomeSourceType.journeyType
+            )
+            )
+          }
+          case (None, _, _) => logAndShowError(isAgent, s"[handleShowRequest]: Could not parse taxYear: $taxYearStringOpt")
+          case (_, None, _) => logAndShowError(isAgent, s"[handleShowRequest]: Could not parse reporting method: $chnageToStringOpt")
+          case (_, _, None) => logAndShowError(isAgent, s"[handleShowRequest]: Could not find incomeSourceId for $incomeSourceType")
+        }
+      )
+    }
+  }
+
+  private def logAndShowError(isAgent: Boolean, errorMessage: String)(implicit user: MtdItUser[_]): Result = {
+    Logger("application").error("[CheckYourAnswersController]" + errorMessage)
+    (if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler).showInternalServerError()
+  }
+
+
+  def submit(isAgent: Boolean, incomeSourceType: IncomeSourceType): Action[AnyContent] = auth.authenticatedAction(isAgent) { implicit user =>
+
+    val successCall = getSuccessCall(isAgent, incomeSourceType)
+    val errorCall = getErrorCall(incomeSourceType, isAgent)
+
+    withSessionData(JourneyType(Manage, incomeSourceType), journeyState = BeforeSubmissionPage) { sessionData =>
+      val (reportingMethodOpt, taxYearOpt, incomeSourceIdStringOpt) = (
```

Comment:
I don't think this data extraction step is required at all

Something like that will do:

```
 withSessionData(JourneyType(Manage, incomeSourceType), journeyState = BeforeSubmissionPage) { sessionData =>
      withIncomeSourcesFS {
        sessionData.manageIncomeSourceData.map(x => (x.reportingMethod, x.taxYear, x.incomeSourceId)) match {
          case Some( (Some(taxYear), Some(reportingMethod), incomeSourceIdStringOpt) ) =>
```